File: backend/app/schemas/shares.py
"""Pydantic schemas for file sharing."""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


# ===========================
# File Share Schemas
# ===========================

class FileShareCreate(BaseModel):
    """Schema for creating a file share with another user."""
    file_id: int = Field(..., description="ID of the file to share")
    shared_with_user_id: int = Field(..., description="User ID to share with")
    can_read: bool = Field(default=True, description="Allow read access")
    can_write: bool = Field(default=False, description="Allow write access")
    can_delete: bool = Field(default=False, description="Allow delete access")
    can_share: bool = Field(default=False, description="Allow re-sharing with others")
    expires_at: Optional[datetime] = Field(default=None, description="Expiration date (None = never expires)")


class FileShareUpdate(BaseModel):
    """Schema for updating a file share."""
    can_read: Optional[bool] = None
    can_write: Optional[bool] = None
    can_delete: Optional[bool] = None
    can_share: Optional[bool] = None
    expires_at: Optional[datetime] = None


class FileShareResponse(BaseModel):
    """Schema for file share response."""
    id: int
    file_id: int
    owner_id: int
    shared_with_user_id: int
    can_read: bool
    can_write: bool
    can_delete: bool
    can_share: bool
    expires_at: Optional[datetime]
    created_at: datetime
    last_accessed_at: Optional[datetime]
    is_expired: bool
    is_accessible: bool
    
    # User information
    owner_username: Optional[str] = None
    shared_with_username: Optional[str] = None
    
    # File information
    file_name: Optional[str] = None
    file_path: Optional[str] = None
    file_size: Optional[int] = None
    is_directory: bool = False

    model_config = {"from_attributes": True}


class SharedWithMeResponse(BaseModel):
    """Schema for files shared with the current user."""
    share_id: int
    file_id: int
    file_name: str
    file_path: str
    file_size: int
    is_directory: bool
    owner_username: str
    owner_id: int
    can_read: bool
    can_write: bool
    can_delete: bool
    can_share: bool
    shared_at: datetime
    expires_at: Optional[datetime]
    is_expired: bool
    
    model_config = {"from_attributes": True}


class ShareStatistics(BaseModel):
    """Statistics for sharing activity."""
    total_file_shares: int
    active_file_shares: int
    files_shared_with_me: int

File: client/src/api/shares.ts
/**
 * API client for file sharing functionality
 */

import { apiClient, apiCache, memoizedApiRequest } from '../lib/api';

export interface FileShare {
  id: number;
  file_id: number;
  owner_id: number;
  shared_with_user_id: number;
  can_read: boolean;
  can_write: boolean;
  can_delete: boolean;
  can_share: boolean;
  expires_at: string | null;
  created_at: string;
  last_accessed_at: string | null;
  is_expired: boolean;
  is_accessible: boolean;
  owner_username: string | null;
  shared_with_username: string | null;
  file_name: string | null;
  file_path: string | null;
  file_size: number | null;
  is_directory: boolean;
}

export interface CreateFileShareRequest {
  file_id: number;
  shared_with_user_id: number;
  can_read?: boolean;
  can_write?: boolean;
  can_delete?: boolean;
  can_share?: boolean;
  expires_at?: string | null;
}

export interface UpdateFileShareRequest {
  can_read?: boolean;
  can_write?: boolean;
  can_delete?: boolean;
  can_share?: boolean;
  expires_at?: string | null;
}

export interface SharedWithMe {
  share_id: number;
  file_id: number;
  file_name: string;
  file_path: string;
  file_size: number;
  is_directory: boolean;
  owner_username: string;
  owner_id: number;
  can_read: boolean;
  can_write: boolean;
  can_delete: boolean;
  can_share: boolean;
  shared_at: string;
  expires_at: string | null;
  is_expired: boolean;
}

export interface ShareStatistics {
  total_file_shares: number;
  active_file_shares: number;
  files_shared_with_me: number;
}

// ===========================
// File Shares API
// ===========================

const SHARES_CACHE_KEY = '/api/shares/user-shares' + JSON.stringify({});

export const createFileShare = async (data: CreateFileShareRequest): Promise<FileShare> => {
  const response = await apiClient.post('/api/shares/user-shares', data);
  apiCache.delete(SHARES_CACHE_KEY);
  return response.data;
};

export const listFileShares = async (): Promise<FileShare[]> => {
  // Memoized: FileShares werden für 60s gecached
  return memoizedApiRequest<FileShare[]>('/api/shares/user-shares');
};

export const listFileSharesForFile = async (fileId: number): Promise<FileShare[]> => {
  const response = await apiClient.get(`/api/shares/user-shares/file/${fileId}`);
  return response.data;
};

export const listFilesSharedWithMe = async (): Promise<SharedWithMe[]> => {
  const response = await apiClient.get('/api/shares/shared-with-me');
  return response.data;
};

export const updateFileShare = async (
  shareId: number,
  data: UpdateFileShareRequest
): Promise<FileShare> => {
  const response = await apiClient.patch(`/api/shares/user-shares/${shareId}`, data);
  apiCache.delete(SHARES_CACHE_KEY);
  return response.data;
};

export const deleteFileShare = async (shareId: number): Promise<void> => {
  await apiClient.delete(`/api/shares/user-shares/${shareId}`);
  apiCache.delete(SHARES_CACHE_KEY);
};

export const getShareStatistics = async (): Promise<ShareStatistics> => {
  const response = await apiClient.get('/api/shares/statistics');
  return response.data;
};
